pkg/proxy/topology.go

```diff
@@ -140,48 +144,77 @@ func CategorizeEndpoints(endpoints []Endpoint, svcInfo ServicePort, nodeLabels m
 	return
 }
 
-// canUseTopology returns true if all of the following are true:
-//   - The node's labels include "topology.kubernetes.io/zone".
-//   - All of the endpoints for this Service have a topology hint.
-//   - At least one endpoint for this Service is hinted for this node's zone.
-func canUseTopology(endpoints []Endpoint, nodeLabels map[string]string) bool {
-	zone, foundZone := nodeLabels[v1.LabelTopologyZone]
+// topologyModeFromHints returns a topology mode ("", "PreferSameZone", or
+// "PreferSameNode") based on the Endpoint hints:
+//   - If the ImprovedTrafficDistribution feature gate is enabled, and every ready
+//     endpoint has a node hint, and at least one endpoint is hinted for this node, then
+//     it returns "PreferSameNode".
+//   - Otherwise, if every ready endpoint has a zone hint, and at least one endpoint is
+//     hinted for this node's zone, then it returns "PreferSameZone".
+//   - Otherwise it returns "" (meaning, no topology / default traffic distribution).
+func topologyModeFromHints(svcInfo ServicePort, endpoints []Endpoint, nodeName, zone string) string {
 	hasEndpointForZone := false
+	allEndpointsHaveZoneHints := true
+	hasEndpointForNode := false
+	allEndpointsHaveNodeHints := true
 	for _, endpoint := range endpoints {
 		if !endpoint.IsReady() {
 			continue
 		}
 
-		// If any of the endpoints do not have zone hints, we bail out
 		if endpoint.ZoneHints().Len() == 0 {
-			klog.V(7).InfoS("Skipping topology aware endpoint filtering since one or more endpoints is missing a zone hint", "endpoint", endpoint)
-			return false
+			allEndpointsHaveZoneHints = false
+		} else if endpoint.ZoneHints().Has(zone) {
+			hasEndpointForZone = true
 		}
 
-		// If we've made it this far, we have endpoints with hints set. Now we check if there is a
-		// zone label, if there isn't one we log a warning and bail out
-		if !foundZone || zone == "" {
-			klog.V(2).InfoS("Skipping topology aware endpoint filtering since node is missing label", "label", v1.LabelTopologyZone)
-			return false
+		if endpoint.NodeHints().Len() == 0 {
+			allEndpointsHaveNodeHints = false
+		} else if endpoint.NodeHints().Has(nodeName) {
+			hasEndpointForNode = true
```

Comment:
Another nit/opinion:
Could this logic move above the ZoneHint logic?

In my head the order is always node, zone, everything else, and it may make the code more readable

Reply:
makes sense